src/libs/elektra/kdb.c

```diff
@@ -1324,6 +1381,441 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
+		}
+	}
+
+	keyDel (mountpointKey);
+	return ret;
+}
+
+enum PluginContractState
+{
+	PLUGIN_STATE_UNMOUNTED,
+	PLUGIN_STATE_MOUNTED,
+	PLUGIN_STATE_REMOUNT,
+};
+
+/**
+ * Ensures a kdbEnsure() contract condition for a global plugin.
+ *
+ * @see kdbEnsure()
+ *
+ * @param handle       the KDB handle
+ * @param pluginName   the name of the plugin
+ * @param pluginState  the intended condition for the plugin
+ * @param pluginConfig the config KeySet for the plugin; is always consumed, i.e. you shouldn't ksDel() it after calling this
+ * @param errorKey     used for error reporting
+ *
+ * @retval 1 on success
+ * @retval 0 otherwise
+ */
+static int ensureGlobalPluginState (KDB * handle, const char * pluginName, enum PluginContractState pluginState, KeySet * pluginConfig,
+				    Key * errorKey)
+{
+	switch (pluginState)
+	{
+	case PLUGIN_STATE_UNMOUNTED:
+		ksDel (pluginConfig);
+		return ensureGlobalPluginUnmounted (handle, pluginName, errorKey);
+	case PLUGIN_STATE_MOUNTED:
+		return ensureGlobalPluginMounted (handle, pluginName, pluginConfig, errorKey);
+	case PLUGIN_STATE_REMOUNT:
+	{
+		return ensureGlobalPluginUnmounted (handle, pluginName, errorKey) &&
+		       ensureGlobalPluginMounted (handle, pluginName, pluginConfig, errorKey);
+	}
+	default:
+		ELEKTRA_ASSERT (0, "missing switch case");
+		return 0;
+	}
+}
+
+/**
+ * Ensures a kdbEnsure() contract condition for a plugin under a certain mountpoint.
+ *
+ * @see kdbEnsure()
+ *
+ * @param handle       the KDB handle
+ * @param mountpoint   the mountpoint to use
+ * @param pluginName   the name of the plugin
+ * @param pluginState  the intended condition for the plugin
+ * @param pluginConfig the config KeySet for the plugin; is always consumed, i.e. you shouldn't ksDel() it after calling this
+ * @param errorKey     used for error reporting
+ *
+ * @retval 1 on success
+ * @retval 0 otherwise
+ */
+static int ensurePluginState (KDB * handle ELEKTRA_UNUSED, const char * mountpoint ELEKTRA_UNUSED, const char * pluginName ELEKTRA_UNUSED,
+			      enum PluginContractState pluginState, KeySet * pluginConfig ELEKTRA_UNUSED, Key * errorKey ELEKTRA_UNUSED)
+{
+	switch (pluginState)
+	{
+	case PLUGIN_STATE_UNMOUNTED:
+		ksDel (pluginConfig);
+		return ensurePluginUnmounted (handle, mountpoint, pluginName, errorKey);
+	case PLUGIN_STATE_MOUNTED:
+		return 0; // TODO: ensurePluginMounted (handle, mountpoint, pluginName, pluginConfig, errorKey);
+	case PLUGIN_STATE_REMOUNT:
+		return 0; // TODO: ensurePluginUnmounted (handle, mountpoint, pluginName, errorKey) && ensurePluginMounted (handle,
+			  // mountpoint, pluginName, pluginConfig, errorKey);
+	default:
+		ELEKTRA_ASSERT (0, "missing switch case");
+		return 0;
+	}
+}
+
+/**
+ * This function can be used the given KDB @p handle meets certain conditions,
+ * specified in @p contract. Currently the following conditions are supported:
+ *
+ * - `system/plugins/<mountpoint>/<pluginname>` defines the state of the plugin
```

Comment:
What about the name `system/elektra/ensure/plugins/...`? `system/plugins` is not reserved and it will create problems once these keys should be read from KDB.

Reply:
I can change it, but I don't see a reason, why these keys should be stored in the KDB. They are not configuration, they are static specification. IMO the use case for `kdbEnsure` is that an application wants to make sure the KDB is in a state that is usable to the application. Changing this contract should only be necessary, if the code changes, i.e. recompiling is necessary.

That is also the reason, why I decided it is okay to consume the contract. If the keys came from the KDB this would be rather inconvenient.

Reply:
> I can change it, but I don't see a reason, why these keys should be stored in the KDB.

It is not about storing them there but it might be useful to be able to query the contracts of applications (with a functionality similar to the specload).